relayer/src/supervisor/scan.rs

```diff
@@ -0,0 +1,749 @@
+
+    fn connection_allowed(
+        &mut self,
+        chain: &Chain,
+        client: &IdentifiedAnyClientState,
+        connection: &IdentifiedConnectionEnd,
+    ) -> bool {
+        if !self.filtering_enabled() {
+            return true;
+        }
+
+        let permission = self.client_state_filter.control_connection_end_and_client(
+            self.registry,
+            &chain.id(),
+            &client.client_state,
+            &connection.connection_end,
+            &connection.connection_id,
+        );
+
+        match permission {
+            Ok(Permission::Deny) => {
+                warn!(
+                    "skipping workers for chain {}, client {} & conn {}, \
+                     reason: client or counterparty client is not allowed",
+                    chain.id(),
+                    client.client_id,
+                    connection.connection_id
+                );
+
+                false
+            }
+            Err(e) => {
+                error!(
+                    "skipping workers for chain {}, client {} & conn {}, reason: {}",
+                    chain.id(),
+                    client.client_id,
+                    connection.connection_id,
+                    e
+                );
+
+                false
+            }
+            _ => true,
+        }
+    }
+
+    fn channel_allowed(&mut self, chain: &Chain, channel: &IdentifiedChannelEnd) -> bool {
+        self.config
+            .packets_on_channel_allowed(&chain.id(), &channel.port_id, &channel.channel_id)
+    }
+}
+
+struct ScannedChannel {
+    channel: IdentifiedChannelEnd,
+    counterparty_channel: Option<IdentifiedChannelEnd>,
+    connection: IdentifiedConnectionEnd,
+    counterparty_connection_state: Option<ConnectionState>,
+    client: IdentifiedAnyClientState,
+}
+
+fn scan_allowed_channel<Chain: ChainHandle>(
+    registry: &'_ mut Registry<Chain>,
+    chain: &Chain,
+    port_id: &PortId,
+    channel_id: &ChannelId,
+) -> Result<ScannedChannel, Error> {
+    let span = info_span!("scan.channel", port_id = %port_id, channel_id = %channel_id);
+    let _guard = span.enter();
+
+    info!("querying channel {}/{}...", port_id, channel_id);
+    let channel = query_channel(chain, port_id, channel_id)?;
+    let connection = query_connection_for_channel(chain, &channel)?;
+    let client_id = connection.connection_end.client_id();
+
+    info!(
+        "found connection {} and client {}",
+        connection.connection_id, client_id
+    );
+
+    info!("querying client {}...", client_id);
+    let client = query_client(chain, client_id)?;
+
+    info!(
+        "client lives on counterparty chain {}",
+        client.client_state.chain_id()
+    );
+
+    let counterparty_chain = registry
+        .get_or_spawn(&client.client_state.chain_id())
+        .map_err(Error::spawn)?;
+
+    let counterparty_channel =
+        channel_on_destination(&channel, &connection, &counterparty_chain).unwrap_or_default();
+
+    info!(
```

Comment:
I think we have a duplicate log line at INFO level: here (line 610) and line 623. Or is this intentional?

I noticed this while starting up a chain, for example see the last log lines here they are redundant:

> 2022-01-19T09:39:53.856979Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}: scanning chain...
2022-01-19T09:39:53.857132Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}: chain uses an allow list, skipping scan for fast startup
2022-01-19T09:39:53.857221Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}: allowed ports/channels: transfer/channel-210
2022-01-19T09:39:53.857348Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}: querying allowed channels...
2022-01-19T09:39:53.857542Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}:scan.channel{port_id=transfer channel_id=channel-210}: querying channel transfer/channel-210...
2022-01-19T09:39:54.120970Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}:scan.channel{port_id=transfer channel_id=channel-210}: found connection connection-368 and client 07-tendermint-419
2022-01-19T09:39:54.121160Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}:scan.channel{port_id=transfer channel_id=channel-210}: querying client 07-tendermint-419...
2022-01-19T09:39:54.209776Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}:scan.channel{port_id=transfer channel_id=channel-210}: client lives on counterparty chain agoricdev-7
2022-01-19T09:39:54.340916Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}:scan.channel{port_id=transfer channel_id=channel-210}: found counterparty channel channel-10
2022-01-19T09:39:54.846280Z  INFO ThreadId(01) scan.chain{chain=vega-testnet}:scan.channel{port_id=transfer channel_id=channel-210}: found counterparty channel channel-10

Reply:
25d5119f